contracts/CollectiveGovernance.sol

```diff
@@ -0,0 +1,106 @@
+ */
+pragma solidity ^0.8.15;
+
+import "./GovernanceStorage.sol";
+import "./Governance.sol";
+import "./UpgradeableGovernance.sol";
+import "./VoteStrategy.sol";
+import "./ElectorVoterPoolStrategy.sol";
+
+/// @title CollectiveGovernance
+// factory contract for governance
+contract CollectiveGovernance is UpgradeableGovernance, Governance {
+    address private owner;
+
+    Storage private _storage;
+    VoteStrategy private _voteStrategy;
+
+    constructor() {
+        owner = msg.sender;
+        _storage = new GovernanceStorage();
+        emit StorageAddress(address(_storage));
+        _voteStrategy = new ElectorVoterPoolStrategy(_storage);
+        emit StrategyChange(0, _voteStrategy.version(), address(_voteStrategy));
+    }
+
+    modifier requireContractOwner() {
+        require(owner == msg.sender, "Not contract owner");
+        _;
+    }
+
+    function setVoteStrategy(address _strategy) external requireContractOwner {
+        uint32 version = _voteStrategy.version();
+        _voteStrategy = VoteStrategy(_strategy);
+        uint32 newVersion = _voteStrategy.version();
+        emit StrategyChange(version, newVersion, _strategy);
+    }
+
+    function getCurrentStrategyVersion() external view returns (uint32) {
+        return _voteStrategy.version();
+    }
+
+    function getCurrentStrategyAddress() external view returns (address) {
+        return address(_voteStrategy);
+    }
+
+    function getStorageAddress() external view returns (address) {
+        return address(_storage);
+    }
+
+    function propose(
+        uint256 quorumThreshold,
+        address erc721,
+        uint256 requiredDuration
+    ) external returns (uint256) {
+        uint256 proposalId = _storage._initializeProposal(address(_voteStrategy));
+        _storage.registerSupervisor(proposalId, msg.sender);
+        _storage.setQuorumThreshold(proposalId, quorumThreshold);
+        _storage.setRequiredVoteDuration(proposalId, requiredDuration);
+        _storage.registerVoterClassERC721(proposalId, erc721);
+        _storage.makeReady(proposalId);
+        _voteStrategy.openVote(proposalId);
+        return proposalId;
+    }
+
+    function endVote(uint256 _proposalId) external {
+        address _strategyAddress = _storage.voteStrategy(_proposalId);
+        VoteStrategy _strategy = VoteStrategy(_strategyAddress);
+        _strategy.endVote(_proposalId);
+    }
+
+    function voteFor(uint256 _proposalId) external {
```

Comment:
what will happen if the strategy changes in the middle of the vote?

Reply:
Actually it looks like the strategy is stored in the proposal so it'll always use the same one, nvm

Reply:
yep - it will fail if someone calls the wrong strategy directly (determined by address) and it will succeed if the address of the strategy contract is the same as the one used to create the proposal